knative-operator/pkg/common/service_monitor.go

```diff
@@ -54,23 +81,210 @@ func SetupServiceMonitor(ctx context.Context, cfg *rest.Config, metricsPort int3
+
+	// If we dont use a custom list here, the typical call to get a filtered list of gvks using k8sutil.GetGVKsFromAddToScheme(apis.AddToScheme)
+	// will end up with resources the Operator does not have access to eg. `Kind=LimitRange`
+	// The list of resources returned is unrelated to our purposes here, thus the customization.
+	gvkFilterList := []schema.GroupVersionKind{
+		schema.GroupVersionKind{
+			Group:   "operator.knative.dev",
+			Version: "v1alpha1",
+			Kind:    "KnativeServing",
+		},
+		schema.GroupVersionKind{
+			Group:   "operator.knative.dev",
+			Version: "v1alpha1",
+			Kind:    "KnativeEventing",
+		},
+	}
+	// To generate metrics in other namespaces, add the values below.
+	// This is due to this bug: https://github.com/operator-framework/operator-sdk/issues/2494
+	// For the workaround check here: https://github.com/operator-framework/operator-sdk/pull/2601/files#r396745465
+	// and https://github.com/shipwright-io/build/pull/73
+	// In order to avoid getting a bad value we avoid using k8sutil.GetWatchNamespace() that gets the value from the WATCH_NAMESPACE
+	// env var. That value is by default "" but user may change it, affecting the metrics endpoint.
+	namespaces := []string{""}
+	// Generate and serve custom resource specific metrics.
+	err := kubemetrics.GenerateAndServeCRMetrics(cfg, namespaces, gvkFilterList, metricsHost, operatorMetricsPort)
 	if err != nil {
 		return err
 	}
-	// Get the namespace the operator is currently deployed in.
-	operatorNs, err := k8sutil.GetOperatorNamespace()
+	return nil
+}
+
+func SetupEventingBrokerServiceMonitors(client client.Client, namespace string, instance *eventingv1alpha1.KnativeEventing) error {
+	if err := isServiceMonitorAvailable(); err != nil {
+		return err
+	}
+	manifest, err := mf.NewManifest(getMonitorPath(TestEventingBrokerServiceMonitorPath, EventingBrokerServiceMonitorPath), mf.UseClient(mfclient.NewClient(client)))
 	if err != nil {
+		return fmt.Errorf("unable to parse broker service monitors: %w", err)
+	}
+	transforms := []mf.Transformer{mf.InjectOwner(instance), mf.InjectNamespace(namespace)}
+	if manifest, err = manifest.Transform(transforms...); err != nil {
+		return fmt.Errorf("unable to transform broker service monitors manifest: %w", err)
+	}
+	// this is required because Apply will fail with not known resource later on
+	monitor := &monitoringv1.ServiceMonitor{}
+	var SchemeGroupVersion = schema.GroupVersion{Group: "monitoring.coreos.com", Version: "v1"}
+	scheme.Scheme.AddKnownTypes(SchemeGroupVersion, monitor)
+	if err := manifest.Apply(); err != nil {
 		return err
 	}
-	// To generate metrics in other namespaces, add the values below.
-	ns := []string{operatorNs}
-	// Generate and serve custom resource specific metrics.
-	err = kubemetrics.GenerateAndServeCRMetrics(cfg, ns, filteredGVK, metricsHost, operatorMetricsPort)
+	return nil
+}
+
+func SetupSourceServiceMonitor(client client.Client, namespace string, instance *appsv1.Deployment) error {
```

Comment:
Does this namespace differ from the namespace in `instance`? If it does, the ownerRef below will break, if it doesn't it's redundant and can be dropped I believe.

Reply:
I followed what the coreos monitoring package does which links the service monitor to service wrt ownership.

Reply:
Regarding the refactoring yes I agree in general will try smaller PRs next. However, it felt to do it in one as the code is simple and is part of one thing that felt to time consuming to break it in let's say 10 PRs. But ok I hear you.

Reply:
@markusthoemmes I will update this shortly.